### ISSUE TITLE
API v4

### ISSUE THREAD

**Issue by SleeplessOne1917** [LemmyNet/lemmy#4428]
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Is your proposal related to a problem?

Since we want to avoid breaking API changes as much as possible after releasing Lemmy v1, I figure the existing API warrants scrutiny. The current REST API design goes against several best practices that many expect to be present in a REST API. For example:

- **Each endpoint should point to a resource.** This is violated by endpoints like GET /post, GET /user, and GET /comment, which all accept the ID for the resource as a query param. Alternatives that follow this best practice would look like GET /post/:id, GET /user/:id, and GET /comment/:id.

- **REST APIs should be stateless.** An example of violations of this are the POST /post/delete, POST /comment/delete, and POST /user/ban. All of these endpoints (and several more like them) rely on the state of the resource: whether or not it is deleted or banned. Each of these accept an argument for whether to set deleted/banned to true or not. To follow best practices, since all of the previously mention things are soft-deletes and revertable bans, we could handle changing these states as part of a PUT or PATCH request to /post/:id.

- **URLs should be nouns (no verbs allowed).** This occurs a lot with Lemmy's API. Just looking at the comments API alone, there is /delete, /remove, /mark\_as\_read, /distinguish, /like, /save, and /report.

### Describe the solution you'd like.

I will come up with a draft for suggested changes. Since this would be a breaking API change, I expect there to be a lot of discussion about this.

### Describe alternatives you've considered.

Accept the v3 API as it is.

**Comment by Nutomic** [comment]
- Lots of actions are under `/user` where they dont belong. Move to `/account/login` or `/moderation/ban` where its more appropriate https://github.com/LemmyNet/lemmy/issues/2022
- Separate image endpoints like `GET /user/{name}/avatar` https://github.com/LemmyNet/lemmy/issues/1772
- https://github.com/LemmyNet/lemmy/issues/4001
- Would be good to fix https://github.com/LemmyNet/lemmy/issues/2444 but that one is very tricky

**Comment by sunaurus** [comment]
I just want to point out that while REST is indeed often portrayed as a "best practice" for designing APIs, it's actually just one way among many of designing things, and was never really intended to be used for HTTP APIs the way it is often done nowadays.

There is nothing inherently wrong with creating an HTTP API which exposes verbs in URLs (`/mark_as_read` etc). Yes, it's not REST, but this is often the most straight-forward and pragmatic way of designing good and easy to understand APIs, and I believe that's really the most important thing.  

Basically, I am saying, does it make sense to actually start refactoring the API like this? It feels like a big effort, both in terms of work in this repo, as well as work on all frontends, and the benefits are not clear for me.

**Comment by dessalines** [comment]
> URLs should be nouns (no verbs allowed). This occurs a lot with Lemmy's API. Just looking at the comments API alone, there is /delete, /remove, /mark_as_read, /distinguish, /like, /save, and /report.

This kinda hints at the limitations of the no-verb rule. Rest only has like 5 verbs and there's no way they could handle all the various intricate actions like save/bookmark, distinguish, ban, remove, etc. Having verbs in the endpoint tell you exactly what its doing. Looking at [reddit's API](https://www.reddit.com/dev/api/), its almost entirely verbs too.

Outside of #2022, I agree that we shouldn't change most of these unless there's a clear benefit.

**Comment by dullbananas** [comment]
https://github.com/LemmyNet/lemmy/issues/2957

**Comment by dullbananas** [comment]
#4459 stores actions for the same type of thing together in the database. Later this can also be done in the API:

* Create structs like `PostActions` with the fields being structs that mostly just have a `published` field

* In database views, use these structs instead of the current flattened style

* Have read/update endpoints for these structs instead of separate endpoints for each action

**Comment by Nutomic** [comment]
As mentioned in https://github.com/LemmyNet/lemmy/pull/4718 and https://github.com/LemmyNet/lemmy/pull/4719, we should use api routes `/account/auth/login`, `/account/auth/register` and `/account/auth/password_reset` with a strict rate limit (ie the register rate limit). That way we dont have to define these endpoints separately, and get rid of the manual implementation for "maximum three password resets per day".

**Comment by SleeplessOne1917** [comment]
I thought of another consideration for this issue. Currently, most endpoints use a mix of DTOs and database models. For example, getting a list of posts returns, among other things, and entire `PostView`. I think it would be a good idea to not return database models and pare down the data that gets returned to only what is needed for each response. This would be simpler than some of the other ideas I've floated by for accomplishing this, such as a sparse REST API or a GrpahQL API.

Thoughts on this?

**Comment by dessalines** [comment]
I believe there's another issue that also suggested GraphQL type query requests. Its probably not worth it, both in terms of coding work, and because people are likely to request the full objects anyway.

**Comment by SleeplessOne1917** [comment]
Do you mean the GraphQL approach isn't worth it, or the paired down DTOs?

**Comment by dessalines** [comment]
GraphQL querying would be overkill. We could remove redundant fields from some of the DB views, as long as those are well-typed and the same everywhere.

**Comment by SleeplessOne1917** [comment]
I could also look into some other approached that take the strengths of GraphQL without some its weaknesses. json api and odata both seem promising.

**Comment by Nutomic** [comment]
See https://github.com/LemmyNet/lemmy/issues/2957

**Comment by Nutomic** [comment]
This is completed.
